File: src/main/java/io/craigmiller160/counter/count/LineCountingTask.java
package io.craigmiller160.counter.count;

import org.apache.commons.lang3.StringUtils;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.concurrent.CountDownLatch;

/**
 * Created by craig on 12/17/16.
 */
public abstract class LineCountingTask implements Runnable {

    private final File file;
    protected final LineCountStorage storage;
    private final CountDownLatch latch;
    private final boolean includeComments;

    protected LineCountingTask(File file, LineCountStorage storage, CountDownLatch latch, boolean includeComments){
        this.file = file;
        this.storage = storage;
        this.latch = latch;
        this.includeComments = includeComments;
    }

    @Override
    public void run() {
        int lineCount = 0;
        try(BufferedReader reader = new BufferedReader(new FileReader(file))){
            String line = null;
            while((line = reader.readLine()) != null){
                if(!StringUtils.isEmpty(line.trim()) && acceptLine(line)){
                    lineCount++;
                }
            }

            addToStorage(lineCount);
        }
        catch(IOException ex){
            ex.printStackTrace();
        }

        latch.countDown();
    }

    public boolean isIncludeComments(){
        return includeComments;
    }

    protected abstract boolean acceptLine(String line);

    protected abstract void addToStorage(int lineCount);
}

File: kotlin-src/io/craigmiller160/kotlin/counter/count/LineCountingTask.kt
package io.craigmiller160.kotlin.counter.count

import java.io.File
import java.util.concurrent.CountDownLatch

abstract class LineCountingTask(val file: File, val storage: LineCountStorage, val latch: CountDownLatch, val includeComments: Boolean) : Runnable

abstract class DefaultLineCountingTask(file: File, storage: LineCountStorage, latch: CountDownLatch, includeComments: Boolean) : LineCountingTask(file, storage, latch, includeComments){

    //TODO any way to make this an abstract property???
    abstract fun getFileType(): String

    override fun run() {
        println("Counting lines in file. Type: ${getFileType()} File: ${file.absolutePath}")

        var lineCount = 0
        file.bufferedReader().useLines { lines ->
            lines.forEach { if(!it.trim().isEmpty() && acceptLine(it)) lineCount++ }
        }

        addToStorage(lineCount)

        latch.countDown()
    }

    //TODO also, acceptLine() should be a function too
    abstract fun acceptLine(line: String): Boolean

    //TODO instead of addToStorage() abstract method, use a passed-in function. Might even need less params here
    abstract fun addToStorage(lineCount: Int)

}